plugins/obs-ffmpeg/ffmpeg-encoded-output.c

```diff
@@ -0,0 +1,915 @@
+{
+	struct ffmpeg_encoded_output *stream =
+	    bzalloc(sizeof(struct ffmpeg_encoded_output));
+
+	stream->output = output;
+	pthread_mutex_init_value(&stream->packets_mutex);
+
+	proto_init(stream);
+
+	if (pthread_mutex_init(&stream->packets_mutex, NULL) != 0)
+		goto fail;
+	if (os_event_init(&stream->stop_event, OS_EVENT_TYPE_MANUAL) != 0)
+		goto fail;
+
+	if (pthread_mutex_init(&stream->write_buf_mutex, NULL) != 0) {
+		warn("Failed to initialize write buffer mutex");
+		goto fail;
+	}
+
+	if (os_event_init(&stream->buffer_space_available_event,
+			  OS_EVENT_TYPE_AUTO) != 0) {
+		warn("Failed to initialize write buffer event");
+		goto fail;
+	}
+	if (os_event_init(&stream->buffer_has_data_event,
+			  OS_EVENT_TYPE_AUTO) != 0) {
+		warn("Failed to initialize data buffer event");
+		goto fail;
+	}
+	if (os_event_init(&stream->send_thread_signaled_exit,
+			  OS_EVENT_TYPE_MANUAL) != 0) {
+		warn("Failed to initialize socket exit event");
+		goto fail;
+	}
+
+	UNUSED_PARAMETER(settings);
+
+	return stream;
+
+ fail:
+	ffmpeg_encoded_output_destroy(stream);
+	return NULL;
+}
+
+static bool ffmpeg_encoded_output_start(void *data)
+{
+	struct ffmpeg_encoded_output *stream = data;
+
+	if (!obs_output_can_begin_data_capture(stream->output, 0))
+		return false;
+	if (!obs_output_initialize_encoders(stream->output, 0))
+		return false;
+
+	os_atomic_set_bool(&stream->connecting, true);
+
+	return pthread_create(&stream->connect_thread, NULL, connect_thread,
+			      stream) == 0;
+}
+
+#define MILLISECOND_DEN   1000
+
+static int32_t get_ms_time(struct encoder_packet *packet, int64_t val)
+{
+	return (int32_t) (val * MILLISECOND_DEN / packet->timebase_den);
+}
+
+static void ffmpeg_encoded_output_data(void *data,
+				       struct encoder_packet *packet)
+{
+	struct ffmpeg_encoded_output *stream = data;
+	struct encoder_packet new_packet;
+	bool added_packet = false;
+
+	if (disconnected(stream) || !active(stream))
+		return;
+
+	if (packet->type == OBS_ENCODER_VIDEO) {
+		if (!stream->got_first_video) {
+			stream->start_dts_offset =
+			    get_ms_time(packet, packet->dts);
+			stream->got_first_video = true;
+		}
+	}
+	obs_encoder_packet_ref(&new_packet, packet);
+
+	pthread_mutex_lock(&stream->packets_mutex);
+
+	if (!disconnected(stream)) {
+		added_packet = (packet->type == OBS_ENCODER_VIDEO) ?
+		    add_video_packet(stream, &new_packet) :
+		    add_packet(stream, &new_packet);
+	}
+
+	pthread_mutex_unlock(&stream->packets_mutex);
+
+	if (added_packet)
+		os_sem_post(stream->send_sem);
+	else
+		obs_encoder_packet_release(&new_packet);
+}
+
+static void ffmpeg_encoded_output_stop(void *data, uint64_t ts)
+{
+	struct ffmpeg_encoded_output *stream = data;
+
+	if (stopping(stream) && ts != 0)
+		return;
+
+	if (connecting(stream))
+		pthread_join(stream->connect_thread, NULL);
+
+	stream->stop_ts = ts / 1000ULL;
+
+	if (ts)
+		stream->shutdown_timeout_ts = ts +
+		    (uint64_t) stream->max_shutdown_time_sec * 1000000000ULL;
```

Comment:
Style:  Tab twice to indent a line continuation.

```suggestion
				(uint64_t) stream->max_shutdown_time_sec *
				1000000000ULL;
```